libr/core/windows_heap.c

```diff
@@ -0,0 +1,822 @@
+		if (block[index].extra & EXTRA_FLAG) {
+			PHeapBlockExtraInfo extra = (PHeapBlockExtraInfo)(block[index].extra & ~EXTRA_FLAG);
+			hb->extraInfo = extra;
+			hb->dwSize -= extra->unusedBytes;
+			hb->dwAddress = (void *)(block[index].address + extra->granularity);
+		} else {
+			hb->extraInfo = NULL;
+			(WPARAM)hb->dwAddress += hb->dwSize;
+		}
+		hb->index++;
+	}
+
+	USHORT flags = block[index].flags;
+	UPDATE_FLAGS (hb, flags);
+
+	return true;
+}
+
+static void free_extra_info(PDEBUG_HEAP_INFORMATION heap) {
+	r_return_if_fail (heap);
+	HeapBlock hb;
+	if (GetFirstHeapBlock (heap, &hb)) {
+		do {
+			R_FREE (hb.extraInfo);
+		} while (GetNextHeapBlock (heap, &hb));
+	}
+}
+
+static bool DecodeHeapEntry(PHEAP heap, PHEAP_ENTRY entry) {
+#if defined(_M_X64)
+	(WPARAM)entry += sizeof (PVOID);
+#endif
+	if (heap->EncodeFlagMask && (*(UINT32 *)entry & heap->EncodeFlagMask)) {
+#if defined(_M_X64)
+		(WPARAM)heap += sizeof (PVOID);
+#endif
+		*(WPARAM *)entry ^= *(WPARAM *)&heap->Encoding;
+	}
+	return !(((BYTE *)entry)[0] ^ ((BYTE *)entry)[1] ^ ((BYTE *)entry)[2] ^ ((BYTE *)entry)[3]);
+}
+
+// Is this right?
+static bool DecodeLFHEntry(PHEAP heap, PHEAP_ENTRY entry, PHEAP_USERDATA_HEADER userBlocks, WPARAM key, WPARAM addr) {
+#if defined(_M_X64)
+	(WPARAM)entry += sizeof (PVOID);
+#endif
+	if (heap->EncodeFlagMask) {
+		*(DWORD *)entry ^= PtrToInt (heap->BaseAddress) ^ (DWORD)(((DWORD)addr - PtrToInt (userBlocks)) << 0xC) ^ (DWORD)key ^ ((DWORD)addr >> 4);
+	}
+	return !(((BYTE *)entry)[0] ^ ((BYTE *)entry)[1] ^ ((BYTE *)entry)[2] ^ ((BYTE *)entry)[3]);
+}
+
+/*
+*	This function may fail with PDI_HEAP_BLOCKS if:
+*		There's too many allocations
+*		The Segment Heap is activated (will block next time called)
+*		Notes:
+*			Some LFH allocations seem misaligned
+*/
+static PDEBUG_BUFFER InitHeapInfo(DWORD pid, DWORD mask) {
+	// Make sure it is not segment heap to avoid lockups
+	if (mask & PDI_HEAP_BLOCKS) {
+		PDEBUG_BUFFER db = InitHeapInfo (pid, PDI_HEAPS);
+		if (db) {
+			HANDLE h_proc = OpenProcess (PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, pid);
+			PHeapInformation heaps = db->HeapInformation;
+			for (int i = 0; i < heaps->count; i++) {
+				DEBUG_HEAP_INFORMATION heap = heaps->heaps[i];
+				if (is_segment_heap (h_proc, heap.Base)) {
+					RtlDestroyQueryDebugBuffer (db);
+					return NULL;
+				}
+			}
+			RtlDestroyQueryDebugBuffer (db);
+		} else {
+			return NULL;
+		}
+	}
+	int res;
+	PDEBUG_BUFFER db = RtlCreateQueryDebugBuffer (0, FALSE);
+	res = RtlQueryProcessDebugInformation (pid, mask, db);
+	if (res) {
+		// why after it fails the first time it blocks on the second? Thats annoying
+		// It stops blocking if i pause radare in the debugger. is it a race?
+		// why it fails with 1000000 allocs? also with processes with segment heap enabled?
+		RtlDestroyQueryDebugBuffer (db);
+		r_sys_perror ("InitHeapInfo");
+		return NULL;
+	}
+	return db;
+}
+
+#define GROW_BLOCKS()										\
```

Comment:
can you line up the `\` so it's more readable? clang-format can do that

Reply:
weird, its lined up in my editor

Reply:
ah, github uses 8 spaces for tabs, your editor is probably using 4